Adds instructions to deploy locally without Cloud Build. (#105)
amandakarina/terraform-example-foundation: 0-bootstrap/README.md

## Running Terraform locally

> The following steps introduce the steps to deploy without Cloud Build.

Reviewer: ```suggestion
The following steps will guide you through deploying without using Cloud Build.
```

1. Clone [terraform-example-foundation](https://github.com/terraform-google-modules/terraform-example-foundation) into your local environment and create to the `gcp-bootstrap` folder at the same level. Copy the `0-bootstrap` content and `.gitignore` to `gcp-bootstrap`.

   ```bash
   git clone https://github.com/terraform-google-modules/terraform-example-foundation.git

   mkdir gcp-bootstrap

   cp -R terraform-example-foundation/0-bootstrap/* gcp-bootstrap/

   cp terraform-example-foundation/.gitignore gcp-bootstrap
   ```

> 1. Navigate to `gcp-bootstrap` and initialize a local git repository, so you can manage versions locally. Create the environment branches.

Reviewer: ```suggestion
1. Navigate to `gcp-bootstrap` and initialize a local Git repository to manage versions locally. Then, Create the environment branches.
```

   ```bash
   cd gcp-bootstrap

   git init
   git checkout -b plan

   git checkout -b shared
   ```

1. Rename `terraform.example.tfvars` to `terraform.tfvars` and update the file with values from your environment:

   ```bash
   mv terraform.example.tfvars terraform.tfvars
   ```

> 1. Rename `cb.tf` to `cb.tf.example` and update the file with values from your environment:

Reviewer: If we are discarding the use of cb.tf then the text should not contain  the phrase "update the file with values from your environment". If that's not the case, ignore the comment

> 1. Rename `cb.tf` to `cb.tf.example` and update the file with values from your environment:

Author: ```suggestion
1. Rename `cb.tf` to `cb.tf.example`:
```

   ```bash
   mv cb.tf cb.tf.example
   ```

> 1. Comment Cloud Build outputs related at `outputs.tf`.

Reviewer: ```suggestion
1. Comment Cloud Build related outputs at `outputs.tf`.
```

optionally, to provide a better UX, you may add the `outputs` names that need to be commented out.

> 1. Comment lines related to Cloud Build at `sa.tf`. Search for `cicd_project_iam_member` and comment the module and the depends_on.

Reviewer: Here is a more clearer version:
```suggestion
1. In `sa.tf` file, comment out lines related to Cloud Build. Specifically, search for `cicd_project_iam_member` and comment out the corresponding module, as well as the "depends_on" meta-argument in any modules that depend on the commented module.
```

1. Use the helper script [validate-requirements.sh](../scripts/validate-requirements.sh) to validate your environment:

   ```bash
   ../terraform-example-foundation/scripts/validate-requirements.sh -o <ORGANIZATION_ID> -b <BILLING_ACCOUNT_ID> -u <END_USER_EMAIL>
   ```

   **Note:** The script is not able to validate if the user is in a Cloud Identity or Google Workspace group with the required roles.

1. Run `terraform init` and `terraform plan` and review the output.

   ```bash
   git checkout plan
   terraform init
   terraform plan -input=false -out bootstrap.tfplan
   ```

> 1. Copy policy repo and copy contents of policy-library to new folder. Create it the folder at the same level of the `terraform-example-foundation` folder.

Reviewer: Here is a reviewed and more clearer suggestion:
```suggestion
1. Create a new folder called gcp-policies at the same directory level as the `terraform-example-foundation` folder. Initialize a Git repository, create a branch called `main`, and copy the contents of the `policy-library` directory from the `terraform-example-foundation` folder into the gcp-policies folder.
```

   ```bash
   cd ../

   mkdir gcp-policies

   cd gcp-policies
   git init
   git checkout -b main
   cp -RT ../terraform-example-foundation/policy-library/ .
   ```

> 1. Commit changes at your main branch to the policy repo. This way you can manage versions locally.

Reviewer: ```suggestion
1. Commit changes to the main branch of the policy repo. This way you can manage versions locally.
```

   ```bash
   git add .
   git commit -m 'Initialize policy library repo'
   ```

1. Navigate out of the policies repo.

   ```bash
   cd ..
   ```

1. To  validate your policies, run `gcloud beta terraform vet`. For installation instructions, see [Install Google Cloud CLI](https://cloud.google.com/docs/terraform/policy-validation/validate-policies#install).

1. Run the following commands and check for violations:

   ```bash
   export VET_PROJECT_ID=pjr-seed-serverless-test
   terraform show -json bootstrap.tfplan > bootstrap.json

>    gcloud beta terraform vet bootstrap.json --policy-library="../terraform-example-foundation/policy-library" --project ${VET_PROJECT_ID}

Reviewer 2: I believe the correct would be:
`gcloud beta terraform vet bootstrap.json --policy-library="./terraform-example-foundation/policy-library" --project ${VET_PROJECT_ID}`

Reviewer: ACK - i'm making this adjustment in another branch and will make a PR to your branch @amandakarina

   ```

   *`A-VALID-PROJECT-ID`* must be an existing project you have access to. This is necessary because `gcloud beta terraform vet` needs to link resources to a valid Google Cloud Platform project.

1. Commit validated code in plan branch.

   ```bash
   git add .
   git commit -m "Initial version os gcp-bootstrap."
   ```

> 1. Checkout `shared` branch and merge plan into it. Run `terraform apply`.

Reviewer: ```suggestion
1. Checkout `shared` branch and merge the `plan` branch into it. Then, Run `terraform apply`.
```

> 1. Commit validated code in plan branch.

Reviewer 2: Missing a `cd gcp-bootstrap`

Reviewer: ACK - i'm making this adjustment in another branch and will make a PR to your branch @amandakarina

   export projects_gcs_bucket_tfstate=$(terraform output -raw projects_gcs_bucket_tfstate)

   echo "network step service account = ${network_step_sa}"
   echo "projects step service account = ${projects_step_sa}"
   echo "projects gcs bucket tfstate = ${projects_gcs_bucket_tfstate}"
   ```

1. Copy the backend and update `backend.tf` with the name of your Google Cloud Storage bucket for Terraform's state. Also update the `backend.tf` of all steps.

   ```bash
   export backend_bucket=$(terraform output -raw gcs_bucket_tfstate)
   echo "backend_bucket = ${backend_bucket}"

   export backend_bucket_projects=$(terraform output -raw projects_gcs_bucket_tfstate)
   echo "backend_bucket_projects = ${backend_bucket_projects}"

   cp backend.tf.example backend.tf

   cd ../

   for i in `find . -name 'backend.tf'`; do sed -i'' -e "s/UPDATE_ME/${backend_bucket}/" $i; done
   for i in `find . -name 'backend.tf'`; do sed -i'' -e "s/UPDATE_PROJECTS_BACKEND/${backend_bucket_projects}/" $i; done

   cd gcp-bootstrap
   ```

1. Re-run `terraform init`. When you're prompted, agree to copy Terraform state to Cloud Storage.

   ```bash
   terraform init
   ```

> 1. Commit the applied code, so you can manage versions locally.

Reviewer: ```suggestion
1. Commit the new code version, so you can manage versions locally.
```

   terraform plan -input=false -out bootstrap.tfplan
   ```

1. Create a new folder called gcp-policies at the same directory level as the `terraform-example-foundation` folder. Initialize a Git repository, create a branch called `main`, and copy the contents of the `policy-library` directory from the `terraform-example-foundation` folder into the gcp-policies folder.

   ```bash
   cd ../

   mkdir gcp-policies

   cd gcp-policies
   git init
   git checkout -b main
   cp -RT ../terraform-example-foundation/policy-library/ .
   ```

1. Commit changes to the main branch of the policy repo. This way you can manage versions locally.

   ```bash
   git add .
   git commit -m 'Initialize policy library repo'
   ```

1. Navigate back to `gcp-bootstrap` repo.

   ```bash
   cd ../gcp-bootstrap
   ```

1. To  validate your policies, run `gcloud beta terraform vet`. For installation instructions, see [Install Google Cloud CLI](https://cloud.google.com/docs/terraform/policy-validation/validate-policies#install).

1. Run the following commands and check for violations:

   ```bash
   export VET_PROJECT_ID=A-VALID-PROJECT-ID
   terraform show -json bootstrap.tfplan > bootstrap.json

>    gcloud beta terraform vet bootstrap.json --policy-library="../terraform-example-foundation/policy-library" --project ${VET_PROJECT_ID}

Reviewer 3: this should use directory `gcp-policies`

Reviewer 3: like `$(pwd)/../gcp-policies`
